pymc3/distributions/bound.py

```diff
@@ -263,47 +75,61 @@ class Bound:
                     'par2', mu=0.0, sigma=1.0, initval=1.0)
     """
 
-    def __init__(self, distribution, lower=None, upper=None):
+    @classmethod
+    def dist(self, distribution, lower=None, upper=None):
         if isinstance(lower, Real) and lower == -np.inf:
             lower = None
         if isinstance(upper, Real) and upper == np.inf:
             upper = None
 
-        if not issubclass(distribution, Distribution):
-            raise ValueError('"distribution" must be a Distribution subclass.')
-
-        self.distribution = distribution
-        self.lower = lower
-        self.upper = upper
-
-    def __call__(self, name, *args, **kwargs):
-        if "observed" in kwargs:
-            raise ValueError(
-                "Observed Bound distributions are not supported. "
-                "If you want to model truncated data "
-                "you can use a pm.Potential in combination "
-                "with the cumulative probability function."
-            )
+        # if "observed" in kwargs:
+        #     raise ValueError(
+        #         "Observed Bound distributions are not supported. "
+        #         "If you want to model truncated data "
+        #         "you can use a pm.Potential in combination "
+        #         "with the cumulative probability function."
+        #     )
 
-        transform = kwargs.pop("transform", "infer")
-        if issubclass(self.distribution, Continuous):
-            return _ContinuousBounded(
-                name, self.distribution, self.lower, self.upper, transform, *args, **kwargs
-            )
-        elif issubclass(self.distribution, Discrete):
-            return _DiscreteBounded(
-                name, self.distribution, self.lower, self.upper, transform, *args, **kwargs
-            )
+        transform = kwargs.pop("transform", None)
+        if issubclass(distribution, Continuous):
+            if transform == None:
+                if lower is not None and upper is not None:
+                    transform = transforms.interval(lower, upper)
+                    default = 0.5 * (lower + upper)
+                elif upper is not None:
+                    transform = transforms.upperbound(upper)
+                else:
+                    transform = transforms.lowerbound(lower)
+
+        elif issubclass(distribution, Discrete):
+            raise ValueError("Can not transform discrete variable.")
         else:
             raise ValueError("Distribution is neither continuous nor discrete.")
 
-    def dist(self, *args, **kwargs):
-        if issubclass(self.distribution, Continuous):
-            return _ContinuousBounded.dist(
-                self.distribution, self.lower, self.upper, *args, **kwargs
-            )
+        if transform is not None:
+            kwargs["transform"] = transform
+        return super.dist([distribution, lower, upper], **kwargs)
```

Comment:
Is distribution a `class` or a `rv_out`? If it's the former this will fail when the argument is passed to the `RandomVariable` no?

Reply:
It's supposed to be an `rv_out`. Maybe we should raise if a class is passed. (Or we could instantiate it?)

Reply:
If it's a rv_out I don't think you can use the `isinstance` class checks above, but the RV op dtype can be used as a proxy. 

I don't think you can instantiate it because you would need to know the arguments, which have not been provided yet. 

Anyway it makes more sense with a rv_out I think. But I wonder if you'll get these rvs bothering you when creating the logp graph...

Reply:
Oh and this is one thing that needs discussion,

Do we keep the previous `pm.Bound` interface of `call` upon a distribution class and then passing it the arguments ? Or do we make the distribution (just the `Distribution.dist`) separately in the Model and then `Bound` it ? 

There is also a middle ground where we can pass the class and arguments of the distribution to `Bound` and it gets combined in the `dist` method of `Bound`.

Reply:
Yeah it needs discussion. Personally I prefer the `.dist` approach as that is much more coherent with the rest of the library.

On the other hand, pm.Bound is being refactored mostly for backwards compatibility because a lot of people have gotten used to it...